hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -2279,3 +2280,58 @@ def attempt(new_ordering):
                     break
             else:
                 i += 1
+
+    def minimize_block_pairs_retaining_sum(self):
+        """This pass minimizes pairs of blocks subject to the constraint that
+        their sum when interpreted as integers remains the same. This allow us
+        to normalize a number of examples that we would otherwise struggle on.
+        e.g. consider the following:
+
+        m = data.draw_bits(8)
+        n = data.draw_bits(8)
+        if m + n >= 256:
+            data.mark_interesting()
+
+        The ideal example for this is m=1, n=255, but we will almost never
+        find that without a pass like this - we would only do so if we
+        happened to draw n=255 by chance.
+
+        This kind of scenario comes up reasonably often in the context of e.g.
+        triggering overflow behaviour.
+        """
+        i = 0
+        while i < len(self.shrink_target.blocks):
+            if self.is_payload_block(i):
+                j = i + 1
+                while j < len(self.shrink_target.blocks):
+                    u, v = self.shrink_target.blocks[i]
+                    m = int_from_bytes(self.shrink_target.buffer[u:v])
+                    if m == 0:
+                        break
+                    r, s = self.shrink_target.blocks[j]
+                    n = int_from_bytes(self.shrink_target.buffer[r:s])
+
+                    if (
+                        s - r == v - u and
+                        self.is_payload_block(j)
+                    ):
+                        def trial(x, y):
+                            if s > len(self.shrink_target.buffer):
+                                return False
+                            attempt = bytearray(self.shrink_target.buffer)
+                            try:
+                                attempt[u:v] = int_to_bytes(x, v - u)
+                                attempt[r:s] = int_to_bytes(y, s - r)
+                            except OverflowError:
+                                return False
+                            return self.incorporate_new_buffer(attempt)
+                        if trial(m - 1, n + 1) and m > 1:
```

Comment:
I think this should be `m >= 1`, so that `m` can be driven all the way to `0`, right?  Also looks like putting the comparison before the call to `trial` might avoid trying to incorporate a bad buffer.  (these probably cancel each other in practice but I'd find it easier to follow the other way)

Reply:
No, this is correct. We know `m > 0`. If `m = 1` we want to run `trial(0, n + 1)` and then stop regardless of what it returns. I'll add a comment to clarify the logic.

Reply:
Aaah, right. That does make sense, but a comment would certainly be good!